src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/MachObjectWriter.cs

```diff
@@ -483,8 +504,10 @@ protected internal override void UpdateSectionAlignment(int sectionIndex, int al
                     Name = name,
                     Section = section,
                     Value = section.VirtualAddress + (ulong)definition.Value,
-                    Descriptor = N_NO_DEAD_STRIP,
-                    Type = N_SECT | N_EXT,
+                    Descriptor =
+                        definition.AltEntry || definition.SectionIndex == _hydrationTargetSectionIndex ?
+                        N_ALT_ENTRY : N_NO_DEAD_STRIP,
```

Comment:
Strictly speaking the `N_NO_DEAD_STRIP` should no longer be required with the changes in this PR. I left it in out of abundance of caution and to minimize the chance of something going wrong.

Reply:
Turns out, caution was warranted. New Xcode linker still messes up the `.hydrated` section even if it's completely composed of `N_ALT_ENTRY` symbols. I changed it to `N_ALT_ENTRY | N_NO_DEAD_STRIP` and I will probably follow up with a bug report to Apple.

Reply:
I found the pattern, will fix ILC to avoid producing it:

test_data.S:
```asm
.subsections_via_symbols
.section __DATA,hydrated
.global .hydrated // <-- symbol we generate at beginning of .hydrated section
.hydrated:
lsection6: // <-- local symbol for beginning of section
.alt_entry _bar_data
.private_extern _foo_data
_foo_data:
  .dword 0
.alt_entry _bar_data
.private_extern _bar_data
_bar_data:
  .dword 1
```

test_code.S:
```
extern char *foo_data;
int main() { return (int)(long)&foo_data; }
```

Build with `clang test_data.S test_code.c -o ./test -Wl,-dead_strip [-Wl,-ld_classic]`. Check output with `nm -nm ./test`

For new linker we get:
```
0000000100000000 (__TEXT,__text) [referenced dynamically] external __mh_execute_header
0000000100003f8c (__TEXT,__text) external _main
0000000100004000 (__DATA,hydrated) external .hydrated
0000000100004000 (__DATA,hydrated) non-external (was a private external) _foo_data
```

For old linker we get:
```
0000000100000000 (__TEXT,__text) [referenced dynamically] external __mh_execute_header
0000000100003f9c (__TEXT,__text) external _main
0000000100004000 (__DATA,hydrated) external .hydrated
0000000100004000 (__DATA,hydrated) non-external (was a private external) _foo_data
0000000100004008 (__DATA,hydrated) non-external (was a private external) _bar_data
```

This only happens if we produce the local section symbol.

Reply:
Filed as Apple bug FB14743667. The workaround in ILC is orthogonal to the changes in this PR, so I will likely submit it separately after I do sufficient testing. Keeping `N_NO_DEAD_STRIP` is sufficient workaround for now.

Reply:
Funnily enough, if I rename `lsectionXX` to `ltmpXX` (ie. the same symbol prefix as emitted for LLVM temporary variables) it works. That seems to be cheap enough workaround.